Add support for directly delivered OTP codes

sbc100 opened the pull request:
Direct OTP codes are ones that are delivered directly to
the user (e.g. SMS) via send_two_factor_authentication_code.
These are randomly generated, short lived, and stored
directly in the database.

TOTP (and the rotp gem) is now only enabled for those user
that have a shared secret (user.create_otp_secret).

sbc100 committed:
Add support for directly delivered OTP codes

Direct OTP codes are ones that are delivered directly to
the user (e.g. SMS) via send_two_factor_authentication_code.
These are randomly generated, short lived, and stored
directly in the database.

TOTP (and the rotp gem) is now only enabled for those user
that have a shared secret (user.create_otp_secret).

sbc100 commented:
@monfresh, what do you think of this change?  I wanted to get your comments before sending anything upstream.

sbc100 commented:
The example app needed the following change:
https://github.com/sbc100/TwoFactorAuthenticationExample/pull/1

monfresh commented on README.md:
```diff
@@ -5,12 +5,15 @@
-* option to encrypt the OTP secret key in the database, with iv and salt
+* Support for 2 types of OTP codes
+ 1. Codes delivered directly to the user
+ 2. TOTP (Google Auenticator) codes based on a shared secret (HMAC)
```
minor spelling error here: Authenticator is misspelled

sbc100 replied:
Done

monfresh commented on app/views/devise/two_factor_authentication/show.html.erb:
```diff
@@ -7,5 +11,9 @@
+<%= link_to "Resend Code", resend_code_user_two_factor_authentication_path, action: :get %>
+<% else %>
+<%= link_to "Send me a code", resend_code_user_two_factor_authentication_path, action: :get %>
+<% end %>
```
Why is the else statement necessary? Resending a code only applies to the direct OTP option, right? If people were sending an SMS to a user while using the TOTP option, they were not using the app properly, so I don't think we should maintain backwards compatibility with that. I'm fine with making this a breaking change and making this version 2.0.

monfresh replied:
Another reason why this is flawed is because this would result in `send_new_otp` to be called in the TOTP scenario, and `send_new_otp` calls `create_direct_otp`, which doesn't apply to the TOTP scenario.

sbc100 replied:
The idea is that TOTP users can always choose to have a code sent to them if they don't have their app.  (all of the TOTP sites I've tried always let you fall back to SMS, for better or for worse).

monfresh replied:
Is the code that is sent during the SMS fallback the actual TOTP code or a one-time direct OTP?

monfresh commented on lib/two_factor_authentication/hooks/two_factor_authenticatable.rb:
```diff
@@ -11,7 +11,9 @@
+      if !user.totp_enabled?
+        user.send_new_otp
+      end
     end
```
The logic here is starting to get unwieldy. I personally don't think this logic belongs here. I think it fits better here: https://github.com/Houdini/two_factor_authentication/blob/6245f9d07cc1dec7e5ae2792db974683d91d4718/lib/two_factor_authentication/controllers/helpers.rb#L16

For now, I suppose this is fine, but this module needs to be refactored at some point.

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -11,17 +11,37 @@ module ClassMethods
+        def authenticate_direct_otp(code, options = {})
+          return false if direct_otp.nil?
+
+          return false if direct_otp != code
```
Can these be combined, like this? `return false if direct_otp.nil? || direct_otp != code`

sbc100 replied:
Done

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -11,17 +11,37 @@ module ClassMethods
+          if Time.now.utc > direct_otp_sent_at + self.class.direct_otp_valid_for
+            # OTP has expired
+            return false
+          end
```
I would put this in a separate method called `direct_otp_not_expired?`, and then the authenticate_direct_otp method can be one line:

``` ruby
direct_otp.present? && direct_otp == code && direct_otp_not_expired?
```

sbc100 replied:
Done (although I whent with direct_otp_expired.. see what you think).

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -11,17 +11,37 @@ module ClassMethods
+          end
+
+          # OTP is valid, but can only be used once so clear it from the DB
+          update_columns(direct_otp: nil, direct_otp_sent_at: nil)
```
I would move this to a separate method called `clear_direct_otp` and call it from the `authenticate_otp` method, like this:

``` ruby
def authenticate_otp(code, options = {})
  if direct_otp and authenticate_direct_otp(code, options)
    clear_direct_otp
    return true
  end
  ...
end
```

sbc100 replied:
Done

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -11,17 +11,37 @@ module ClassMethods
+        end
+
+        def authenticate_totp(code, options = {})
+          raise "authenticate_totp called with no otp_secret_key set" if otp_secret_key.nil?
```
I don't think this is needed because the only time `authenticate_totp` will be called is via the `authenticate_otp` method in this line:

```
return true if totp_enabled? and authenticate_totp(code, options)
```

If `totp_enabled?` returns false, `authenticate_totp` will never get called, so it's impossible for it to get called if otp_secret_key is nil

sbc100 replied:
I think its useful to users to be able to call authenticate_totp or authenticate_direct_otp directly rather than the hybrid function that basically allows for both.   Some application might want to be clear about exactly what type of code the user is trying to use (e.g. Google does this.. it won't allow you to type a direct OTP in into the TOTP dialog box).

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -30,14 +50,15 @@ def authenticate_otp(code, options = {})
         end
 
-        def otp_code(time = Time.now, options = {})
+        def generate_totp_code(time = Time.now, options = {})
```
I don't see this method being used anywhere. Unless I missed it, I couldn't find a test for the Google Authenticator scenario. I would assume that a controller or feature spec would call this method to allow the user to enter the code from their Authenticator app.

Also, I would call this method `totp_code`, not `generate_totp_code` because what this method returns is the user's TOTP code at the current time. It does not generate a new code.

sbc100 replied:
This method is only used for testing.  I moved it into a spec helper class.

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -30,14 +50,15 @@ def authenticate_otp(code, options = {})
 
-        def otp_code(time = Time.now, options = {})
+        def generate_totp_code(time = Time.now, options = {})
+          raise "generate_totp_code called with no otp_secret_key set" if otp_secret_key.nil?
```
I don't think this is necessary. This method would only be called in tests, so it's up to the tests to make sure the user is totp_enabled. If anything, I would remove this method from this module and just make it available as a test helper because it should never be called in the app.

monfresh replied:
The one reason we would want to leave this method here is to support apps that were sending the TOTP code via SMS or some other method, but I would rather not support bad behavior.

sbc100 replied:
Done.

monfresh commented:
It feels a little incomplete to me that we are separating TOTP from direct OTP without providing a way in the UI to enable TOTP for a user. Having the `create_otp_secret` method by itself doesn't really help because it would only make sense to run that method if the user chose to enable 2FA via an Authenticator app, or if you were building an app that only supports TOTP.

I propose that we either implement a basic 2FA setup screen that allows a user to choose between direct OTP and TOTP, or update the README to explain that this exercise is left up to the user, but that it's necessary if a user is to be allowed to choose between both methods.

monfresh commented:
In the case where an app only supports TOTP, the README should suggest that `create_otp_secret` should be called for all existing users, and via a `before_create` callback in the User model for new users, just like it used to be.

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -30,14 +50,15 @@ def authenticate_otp(code, options = {})
         end
 
         def provisioning_uri(account = nil, options = {})
+          raise "provisioning_uri called with no otp_secret_key set" if otp_secret_key.nil?
```
I think Rubocop would complain about this in that the first argument should be an error class. I think it's preferable to follow the existing pattern as in the `send_two_factor_authentication_code` method:

``` ruby
raise NotImplementedError.new("No default implementation - please define in your class.")
```

Also, I would add more details to the message to explain that `create_otp_secret` needs to be called to set the `otp_secret_key`.

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -46,6 +67,18 @@ def need_two_factor_authentication?(request)
         end
 
+        def otp_code
+          direct_otp
```
If we want to maintain backwards compatibility, this should point to `totp_code` (i.e. `generate_totp_code` as you have it named). In the specs in this PR that test the direct OTP scenario, they would call `user.direct_otp`, not `user.otp_code`.

sbc100 replied:
I left otp_code here only for backward compat.   If you are ok with breaking that I will just drop this method completely.

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -58,9 +91,25 @@ def max_login_attempts
 
-        def populate_otp_column
+        def totp_enabled?
+          self.respond_to?(:otp_secret_key) && !self.otp_secret_key.nil?
```
I believe the `self` here is not necessary.

sbc100 replied:
Done.  self.respond_to? is used elsewhere in this file, so I updated that too.

monfresh commented on lib/two_factor_authentication/models/two_factor_authenticatable.rb:
```diff
@@ -58,9 +91,25 @@ def max_login_attempts
+        def create_direct_otp(options = {})
+          # Create a new random OTP and store it in the database
+          digits = options[:direct_otp_length] || self.class.direct_otp_length
+          update_columns(direct_otp: random_base10(digits), direct_otp_sent_at: Time.now.utc)
```
I believe the recommendation is to use `Time.current` instead of `Time.now`. See https://robots.thoughtbot.com/its-about-time-zones#in-summary

sbc100 replied:
But I'm converting directly to UTC, so shouldn't Time.now.utc so it doesn't matter here does it?  System time should be fine.

monfresh commented on lib/two_factor_authentication/version.rb:
```diff
@@ -1,3 +1,3 @@
 module TwoFactorAuthentication
-  VERSION = "1.1.5".freeze
+  VERSION = "1.1.6".freeze
 end
```
The version number update is usually left to the maintainer. Not every PR might merit a version update. Some people like to combine several contributions into a version update.

sbc100 replied:
Done

monfresh commented on spec/features/two_factor_authenticatable_spec.rb:
```diff
@@ -18,12 +18,12 @@
         message = SMSProvider.last_message
         expect(message.to).to eq(user.phone_number)
-        expect(message.body).to eq(user.otp_code)
+        expect(message.body.length).to eq(6)
```
Don't we also want to test that the correct OTP code was sent to the user, not just the length?

sbc100 replied:
I will fix this.  The problem is that accessing the otp_code now requires a reload, since its stored in the database.   For some reason adding a reload here causes other tests to fail.

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -44,33 +44,76 @@
+      expect(instance.otp_code).to be_nil
+      instance.create_direct_otp
+      expect(instance.otp_code).not_to be_nil
+    end
```
Since the `create_direct_otp` method updates the `direct_otp` column in the DB, I would expect this test to make sure the `direct_otp` column was updated, as opposed to indirectly checking it via `otp_code`.

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -44,33 +44,76 @@
+      instance.create_direct_otp
+      expect(instance.otp_code.length).to equal(6)
+    end
+  end
```
Same comment here. Replace `otp_code` with `direct_otp` to make it clearer what is being tested.

monfresh replied:
I would also add a test to verify that you can pass in the digits as a parameter to the `create_direct_otp` method.

monfresh replied:
And another test that makes sure that the `direct_otp_sent_at` column is set to the current time.

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -44,33 +44,76 @@
+      end
+    end
+  end
+
```
I would also add a test for the scenario where a direct OTP exists, but the one passed in to the method does not match.

sbc100 replied:
Done

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -90,66 +133,67 @@ def instance.send_two_factor_authentication_code
-      end
+        before do
+          instance.email = 'houdini@example.com'
+          instance.run_callbacks :create
```
I don't think we need this anymore now that you have removed the `before_create` callback.

sbc100 replied:
Done

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -90,66 +133,67 @@ def instance.send_two_factor_authentication_code
-
-      it 'repopulates otp_column' do
+      it 'populates otp_secret_key column' do
         instance.run_callbacks :create
```
This is no longer needed.

sbc100 replied:
Done

monfresh commented on spec/rails_app/config/initializers/devise.rb:
```diff
@@ -142,7 +142,7 @@
   # one (and only one) @ exists in the given string. This is mainly
   # to give user feedback and not to assert the e-mail validity.
-  # config.email_regexp = /\A[^@]+@[^@]+\z/
+  config.email_regexp = /\A[^@\s]+@[^@\s]+\z/
```
Why is it necessary to turn this on for this PR?

sbc100 replied:
Removed a deprecation warning.. but not part of this PR.  Reverted.

monfresh commented:
Another thing you'll need to do is create a migration to add the new direct OTP columns, then run `rake db:migrate` from within the `spec/rails_app` folder to update the schema for the test Rails app that is used by the specs. See this file as an example from my [PR](https://github.com/Houdini/two_factor_authentication/pull/62/files):

spec/rails_app/db/migrate/20151224171231_add_encrypted_columns_to_user.rb

If we want the direct OTP DB columns to be added automatically when someone new uses this gem, we'll also need to update these two files:

lib/generators/active_record/templates/migration.rb
lib/two_factor_authentication/schema.rb

and we'll need to update the README to let existing users know that they if they want to support direct OTP, they need to add two new columns.

monfresh commented on README.md:
```diff
@@ -113,17 +116,22 @@ In the example above, two factor authentication will not be required for local
+may retrieve a one-time password directly from the Google Authenticator app.
 
 #### Overriding the view
 
```
We should also update the `Automatic initial setup` and `Manual initial setup` sections to mention the new DB columns.

sbc100 commented:
I think I've addressed most of your comments.

Regarding creating migration for spec/rails_app... it seems like its working OK without this, which makes a me a little suspicious  Do many tests actually use that app?

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -2,154 +2,172 @@
       before :each do
         instance.otp_secret_key = '2z6hxkdwi3uvrnpn'
+        @totp_helper = TotpHelper.new(instance.otp_secret_key, instance.class.otp_length)
       end
```
Why is the helper needed here? It wasn't needed before in these tests. What changed that causes these tests to require a helper?

sbc100 replied:
Before: We needed to know the current TOTP value at runtime so we could deliver it to the user.
Now: We don't need to know the TOTP values at runtime at all.. we only need to verify the ones that user enters.

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -2,154 +2,172 @@
 
-    it_behaves_like 'otp_code', GuestUser.new
-    it_behaves_like 'otp_code', EncryptedUser.new
+      it 'revents code re-use' do
```
minor spelling: revents -> prevents

sbc100 replied:
Done

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -2,154 +2,172 @@
-        context 'with a known time' do
-          let(:time) { 1_392_852_756 }
+    it "honors 'direct_otp_length' in options paramater" do
+      instance.create_direct_otp({length: 8})
```
Rubocop would say that the brackets are redundant here (and a couple lines below as well).

sbc100 replied:
Done

monfresh commented on spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb:
```diff
@@ -2,154 +2,172 @@
-          instance.otp_secret_key = '2z6hxkdwi3uvrnpn'
-        end
+    it 'set direct_otp_send_at field to current time' do
+      Timecop.freeze()
```
When freezing time outside of a block, `Timecop.return` must always be called at the end of the test to avoid any time-based issues with subsequent tests. Alternatively, always use a block with Timecop.freeze, such as:

``` ruby
Timecop.freeze(Time.now) do
  instance.create_direct_otp
  expect(instance.direct_otp_sent_at).to eq(Time.now)
end
```

See https://github.com/travisjeffery/timecop#timecopsafe_mode

sbc100 replied:
Done

sbc100 committed:
Seperate totp secret generation from confirmation

For most use cases the totp secret needs to be transmitted
to the end user so it its helpful to be able to generate
it, before confirming it.
